File: ios/ArmyLight/Protocol/Palette.swift
//
//  Palette.swift
//  ARMY Light
//
//  The color palette and named-color lookup — a Swift port of the Python
//  `army_light/palette.py`. Menu order is preserved so the iOS grid matches
//  the macOS menu.
//

import Foundation
import SwiftUI

/// A labeled palette color.
struct PaletteColor: Identifiable, Hashable {
    let label: String
    let rgb: RGB
    var id: String { label }
}

enum Palette {
    /// Grid order, top to bottom / left to right.
    static let all: [PaletteColor] = [
        PaletteColor(label: "Red",    rgb: RGB(255, 0, 0)),
        PaletteColor(label: "Orange", rgb: RGB(255, 80, 0)),
        PaletteColor(label: "Yellow", rgb: RGB(255, 200, 0)),
        PaletteColor(label: "Green",  rgb: RGB(0, 255, 0)),
        PaletteColor(label: "Cyan",   rgb: RGB(0, 200, 255)),
        PaletteColor(label: "Blue",   rgb: RGB(0, 0, 255)),
        PaletteColor(label: "Purple", rgb: RGB(160, 0, 255)),
        PaletteColor(label: "Pink",   rgb: RGB(255, 40, 150)),
        PaletteColor(label: "White",  rgb: RGB(255, 255, 255)),
        PaletteColor(label: "Off",    rgb: RGB(0, 0, 0)),
    ]

    /// Lowercased name -> rgb.
    static let byName: [String: RGB] = Dictionary(
        uniqueKeysWithValues: all.map { ($0.label.lowercased(), $0.rgb) }
    )

    /// Accept a palette name ("red"), "r,g,b", or "#rrggbb". Returns nil if unparseable.
    static func parse(_ text: String) -> RGB? {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if let named = byName[t] { return named }
        if t.hasPrefix("#"), t.count == 7 {
            let hex = t.dropFirst()
            let s = hex.startIndex
            guard
                let r = Int(hex[s..<hex.index(s, offsetBy: 2)], radix: 16),
                let g = Int(hex[hex.index(s, offsetBy: 2)..<hex.index(s, offsetBy: 4)], radix: 16),
                let b = Int(hex[hex.index(s, offsetBy: 4)..<hex.index(s, offsetBy: 6)], radix: 16)
            else { return nil }
            return RGB(r, g, b)
        }
        if t.contains(",") {
            let parts = t.split(separator: ",").map { Int($0.trimmingCharacters(in: .whitespaces)) }
            if parts.count == 3, parts.allSatisfy({ $0 != nil && (0...255).contains($0!) }) {
                return RGB(parts[0]!, parts[1]!, parts[2]!)
            }
        }
        return nil
    }
}

// MARK: - SwiftUI bridging

extension RGB {
    /// A SwiftUI Color for this triple.
    var color: Color {
        Color(red: Double(r) / 255, green: Double(g) / 255, blue: Double(b) / 255)
    }

    /// "#RRGGBB" uppercase.
    var hexString: String { String(format: "#%02X%02X%02X", r, g, b) }

    /// Build from a SwiftUI Color (best-effort via its resolved RGB).
    init(color: Color) {
        let resolved = color.resolve(in: EnvironmentValues())
        self.init(
            Int((resolved.red * 255).rounded()),
            Int((resolved.green * 255).rounded()),
            Int((resolved.blue * 255).rounded())
        )
    }
}

File: army_light/palette.py
"""The menu color palette and named-color lookup."""

from __future__ import annotations

RGB = tuple[int, int, int]

# Menu order, top to bottom. (label, (r, g, b)). "Off" renders as its own
# menu section, not in the COLORS grid.
PALETTE: list[tuple[str, RGB]] = [
    ("Red", (255, 0, 0)),
    ("Orange", (255, 80, 0)),
    ("Amber", (255, 150, 0)),
    ("Yellow", (255, 210, 0)),
    ("Lime", (160, 255, 0)),
    ("Green", (0, 255, 0)),
    ("Mint", (0, 255, 150)),
    ("Cyan", (0, 200, 255)),
    ("Sky", (80, 140, 255)),
    ("Blue", (0, 0, 255)),
    ("ARMY Purple", (130, 60, 255)),
    ("Purple", (160, 0, 255)),
    ("Magenta", (255, 0, 255)),
    ("Pink", (255, 40, 150)),
    ("Rose", (255, 120, 170)),
    ("White", (255, 255, 255)),
    ("Off", (0, 0, 0)),
]

# Lowercased name -> rgb, for `probe --color red` etc.
BY_NAME: dict[str, RGB] = {label.lower(): rgb for label, rgb in PALETTE}


def parse_color(text: str) -> RGB:
    """Accept a palette name ('red'), 'r,g,b', or '#rrggbb'."""
    t = text.strip().lower()
    if t in BY_NAME:
        return BY_NAME[t]
    if t.startswith("#") and len(t) == 7:
        return (int(t[1:3], 16), int(t[3:5], 16), int(t[5:7], 16))
    if "," in t:
        parts = [int(p) for p in t.split(",")]
        if len(parts) == 3 and all(0 <= p <= 255 for p in parts):
            return (parts[0], parts[1], parts[2])
    raise ValueError(f"Can't parse color {text!r} (try 'red', '255,0,0', or '#ff0000').")
